rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/CoordinatorBookKeeper.java

```diff
@@ -0,0 +1,38 @@
+/**
+ * Copyright (c) 2016. Qubole Inc
+ * Licensed under the Apache License, Version 2.0 (the License);
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License. See accompanying LICENSE file.
+ */
+
+package com.qubole.rubix.bookkeeper;
+
+import com.codahale.metrics.MetricRegistry;
+import com.qubole.rubix.bookkeeper.manager.CoordinatorManager;
+import org.apache.hadoop.conf.Configuration;
+
+import java.io.FileNotFoundException;
+
+public class CoordinatorBookKeeper extends BookKeeper
+{
+  // The manager used when running on a coordinator node.
+  private final CoordinatorManager coordinatorManager;
+
+  public CoordinatorBookKeeper(Configuration conf, MetricRegistry metrics, CoordinatorManager coordinatorManager) throws FileNotFoundException
+  {
+    super(conf, metrics);
+    this.coordinatorManager = coordinatorManager;
+  }
+
+  @Override
+  public void handleHeartbeat(String workerHostname)
```

Comment:
Sorry - I am still confused by the flow. Who call this function ?

Reply:
* `HeartbeatService` creates a client to connect the master BookKeeper daemon.
* `HeartbeatService` periodically calls `handleHeartbeat` on the client to send the heartbeat to the master via Thrift.
* This `handleHeartbeat` is called when the Thrift call reaches the master.